contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,174 @@
+     3- Calculate gains/loss
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        // reads
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        // interactions
+        //Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
+            IERC20(LPtoken).approve(masterChef, balance);
+            IMasterChef(masterChef).deposit(id, balance); // deposit into Onsen
+        }
+
+        // Stake the sushi tokens
+        if (IERC20(sushi).balanceOf(address(this)) > 0) {
+            enterSushiBar(true); // manage rewards
+        }
+
+        //Calculate gains/loss
+        // Retrieve current balance for pool and address
+        UserInfo memory currentUserInfo = IMasterChef(masterChef).userInfo(id, address(this));
+
+        uint256 last = extender.getAllocatorAllocated(id) + extender.getAllocatorPerformance(id).gain;
+
+        if (currentUserInfo.amount >= last) {
+            gain = uint128(currentUserInfo.amount - last);
+        } else {
+            loss = uint128(last - currentUserInfo.amount);
+        }
+    }
+
+    /**
+        Asuming the index of the amounts [] can be use as parameter for the mapping tokenIds
+        Example:
+        amounts => [20,50,150]
+        So the values that will be used to retrieve the id used in onsen will be 0, 1 , 2
+    */
+    function deallocate(uint256[] memory amounts) public override onlyGuardian {
+        for (uint256 i; i < amounts.length; i++) {
+            uint256 amount = amounts[i];
+            if (amount > 0) {
+                IMasterChef(masterChef).withdraw(tokenIds[i], amount); // withdraw from Onsen
+            }
+        }
+    }
+
+    function _deactivate(bool panic) internal override {}
+
+    function _prepareMigration() internal override {}
+
+    function amountAllocated(uint256 id) public view override returns (uint256) {
+        UserInfo memory currentUserInfo = IMasterChef(masterChef).userInfo(id, address(this));
+        return currentUserInfo.amount;
+    }
+
+    function rewardTokens() public view override returns (IERC20[] memory) {
+        IERC20[] memory rewards = new IERC20[](1);
+        rewards[0] = IERC20(sushi);
+        return rewards;
+    }
+
+    function utilityTokens() public view override returns (IERC20[] memory) {
+        IERC20[] memory utility = new IERC20[](2);
+        utility[0] = IERC20(sushi);
+        utility[1] = IERC20(xSushi);
+        return utility;
+    }
+
+    function name() external pure override returns (string memory) {
+        return "OnsenAllocator";
+    }
+
+    /* ========== INTERNAL FUNCTIONS ========== */
+
+    /**
+     * @notice stake sushi rewards if enter is true else return funds to treasury.
+     * @param _stake bool
+     */
+    function enterSushiBar(bool _stake) internal {
```

Comment:
Also, why isthis internal? The only time you call this is in update and pass in true. So you would never be able to unstake?

Also, the naming does not really work. if  `_stake` is false, you aren't entering the sushi bar, you're exiting.

Reply:
> rename to `_enterSushiBar` to denote an internal function
A/ I used this name because it reflects the contract and the specific function that is being called from ONSEN/SUSHI
SushiBar.sol/function enter 
But I can change it to:
function manageSushi(bool ) internal {

Reply:
> Also, why isthis internal? The only time you call this is in update and pass in true. So you would never be able to unstake?
> 
> Also, the naming does not really work. if `_stake` is false, you aren't entering the sushi bar, you're exiting.

This is internal because it is handling what to do with sushi token (rewards), once a LP is deposited into ONSEN sushi tokens are transferred to the sender.
The logic behind the _stake was that you could send the sushi tokens back to the treasury or stake them, so the idea of this function was not to un-stake, I will change the function to avoid passing a boolean and always just stake the sushi tokens.

Reply:
> rename to `_enterSushiBar` to denote an internal function

I used this name because it reflects the contract and the specific function that is being called from ONSEN/SUSHI
SushiBar.sol/function enter
But I can change it to something like:
function stakeSushi() internal {

Since this function is only going to be doing that.

Reply:
Changes applied:
function enterSushiBar(bool _stake) internal {
to
function stakeSushi() internal {

only function is to stake sushi tokens and send to treasury xSushi tokens